packages/components/src/scroll-lock/index.js

```diff
@@ -1,114 +1,74 @@
-	 * as preventing default on certain touchmove events.
-	 */
-	let previousScrollTop = 0;
+function useSetLocked( className ) {
+	/** @type {import('react').MutableRefObject<number>} */
+	const previousScrollTop = useRef( 0 );
 
 	/**
-	 * Locks and unlocks scroll depending on the boolean argument.
-	 *
-	 * @param {boolean} locked Whether or not scroll should be locked.
+	 * @param {boolean} locked
 	 */
 	function setLocked( locked ) {
-		const scrollingElement =
-			htmlDocument.scrollingElement || htmlDocument.body;
+		const scrollingElement = document.scrollingElement || document.body;
 
 		if ( locked ) {
-			previousScrollTop = scrollingElement.scrollTop;
+			previousScrollTop.current = scrollingElement.scrollTop;
 		}
 
 		const methodName = locked ? 'add' : 'remove';
 		scrollingElement.classList[ methodName ]( className );
 
 		// Adding the class to the document element seems to be necessary in iOS.
-		htmlDocument.documentElement.classList[ methodName ]( className );
+		document.documentElement.classList[ methodName ]( className );
 
 		if ( ! locked ) {
-			scrollingElement.scrollTop = previousScrollTop;
+			scrollingElement.scrollTop = previousScrollTop.current;
 		}
 	}
 
-	/**
-	 * Requests scroll lock.
-	 *
-	 * This function tracks requests for scroll lock. It locks scroll on the first
-	 * request and counts each request so `releaseLock` can unlock scroll when
-	 * all requests have been released.
-	 */
-	function requestLock() {
+	return setLocked;
+}
+
+/*
+ * Setting `overflow: hidden` on html and body elements resets body scroll in iOS.
+ * Save scroll top so we can restore it after locking scroll.
+ *
+ * NOTE: It would be cleaner and possibly safer to find a localized solution such
+ * as preventing default on certain touchmove events.
+ */
+let lockCounter = 0;
+
+/**
+ * A component that will lock scrolling when it is mounted and unlock scrolling when it is unmounted.
+ *
+ * @param {Object} props
+ * @param {string} [props.className]
+ * @return {null} Render nothing.
+ */
+export default function ScrollLock( { className = 'lockscroll' } ) {
+	const setLocked = useSetLocked( className );
+
+	useEffect( () => {
 		if ( lockCounter === 0 ) {
 			setLocked( true );
 		}
 
 		++lockCounter;
-	}
 
-	/**
-	 * Releases a request for scroll lock.
-	 *
-	 * This function tracks released requests for scroll lock. When all requests
-	 * have been released, it unlocks scroll.
-	 */
-	function releaseLock() {
-		if ( lockCounter === 1 ) {
-			setLocked( false );
-		}
+		return () => {
+			if ( lockCounter === 1 ) {
+				setLocked( false );
+			}
 
-		--lockCounter;
-	}
-
-	return class ScrollLock extends Component {
-		/**
-		 * Requests scroll lock on mount.
-		 */
-		componentDidMount() {
-			requestLock();
-		}
-		/**
-		 * Releases scroll lock before unmount.
-		 */
-		componentWillUnmount() {
-			releaseLock();
-		}
+			--lockCounter;
+		};
+	}, [] );
```

Comment:
The `react-hooks/exhaustive-deps` rule is going to complain about missing `setLocked` here.

Reply:
And now it's going to complain about the `className` prop.

Specifying `className` as prop is not entirely correct. All instances of `<ScrollLock />` should share one `className`, because they share the global lock state. Multiple instances with different `className`s will lead to broken CSS classes on the `document.body` element. The class name wasn't really customizable in the original version, except tests.